src/nodelet/adding_image_nodelet.cpp

```diff
@@ -0,0 +1,155 @@
+    Config config_;
+    boost::shared_ptr<ReconfigureServer> reconfigure_server_;
+
+    image_transport::SubscriberFilter sub_image1_, sub_image2_;
+    image_transport::Publisher img_pub_;
+    boost::shared_ptr<message_filters::Synchronizer<SyncPolicy> > sync_;
+    boost::shared_ptr<message_filters::Synchronizer<ApproxSyncPolicy> > async_;
+    boost::mutex mutex_;
+
+    bool approximate_sync_;
+    double alpha_;
+    double beta_;
+    double gamma_;
+
+    void subscribe() {
+      NODELET_DEBUG("Subscribing to image topic.");
+      sub_image1_.subscribe(*it_, "image1", 1);
+      sub_image2_.subscribe(*it_, "image2", 1);
+      if (approximate_sync_) {
+        async_ =
+          boost::make_shared<message_filters::Synchronizer<ApproxSyncPolicy> >(
+            100);
+        async_->connectInput(sub_image1_, sub_image2_);
+        async_->registerCallback(
+          boost::bind(&AddingImageNodelet::add, this, _1, _2));
+      } else {
+        sync_ =
+          boost::make_shared<message_filters::Synchronizer<SyncPolicy> >(100);
+        sync_->connectInput(sub_image1_, sub_image2_);
+        sync_->registerCallback(
+          boost::bind(&AddingImageNodelet::add, this, _1, _2));
+      }
+    }
+
+    void unsubscribe() {
+      sub_image1_.unsubscribe();
+      sub_image2_.unsubscribe();
+    }
+
+    void reconfigureCallback(Config& config, uint32_t level) {
+      boost::mutex::scoped_lock lock(mutex_);
+      config_ = config;
+      alpha_ = config.alpha;
+      beta_ = config.beta;
+      gamma_ = config.gamma;
+    }
+
+    void add(const sensor_msgs::Image::ConstPtr& image_msg1,
+             const sensor_msgs::Image::ConstPtr& image_msg2) {
+      cv::Mat image1 =
+        cv_bridge::toCvShare(image_msg1, image_msg1->encoding)->image;
+      cv::Mat image2 =
+        cv_bridge::toCvShare(image_msg2, image_msg2->encoding)->image;
+      cv::Mat result_image;
+      cv::addWeighted(image1, alpha_, image2, beta_, gamma_, result_image);
+      img_pub_.publish(cv_bridge::CvImage(image_msg1->header,
+                                          image_msg1->encoding,
+                                          result_image).toImageMsg());
+    }
+
+  public:
+    virtual void onInit() {
+      Nodelet::onInit();
+      it_ = boost::shared_ptr<image_transport::ImageTransport>(
+        new image_transport::ImageTransport(*nh_));
+
+      ////////////////////////////////////////////////////////
+      // Dynamic Reconfigure
+      ////////////////////////////////////////////////////////
+      reconfigure_server_ =
+        boost::make_shared<dynamic_reconfigure::Server<Config> >(*pnh_);
+      dynamic_reconfigure::Server<Config>::CallbackType f =
+        boost::bind(&AddingImageNodelet::reconfigureCallback, this, _1, _2);
+      reconfigure_server_->setCallback(f);
+
+      pnh_->param("approximate_sync", approximate_sync_, true);
+      img_pub_ = advertiseImage(*pnh_, "image", 1);
```

Comment:
1. this code is basically copied from sample code, so we do not have to
remove code, even if that is unnecessary
2. that gui will be sometimes good debug tools, if you can not see in the
gui, there is something wrong on input, if you can see the result in gui,
but not seen in output image_view, then something wrong after opencv_apps
nodes output

--
◉ Kei Okada


On Wed, Aug 31, 2016 at 3:29 PM, iory <notifications@github.com> wrote:

> In src/nodelet/adding_image_nodelet.cpp
> <https://github.com/ros-perception/opencv_apps/pull/29#discussion_r76930698>
> :
>
> > +    virtual void onInit() {
> > +      Nodelet::onInit();
> > +      it_ = boost::shared_ptr<image_transport::ImageTransport>(
> > +        new image_transport::ImageTransport(*nh_));
> > +
> > +      ////////////////////////////////////////////////////////
> > +      // Dynamic Reconfigure
> > +      ////////////////////////////////////////////////////////
> > +      reconfigure_server_ =
> > +        boost::make_shared<dynamic_reconfigure::Server<Config> >(*pnh_);
> > +      dynamic_reconfigure::Server<Config>::CallbackType f =
> > +        boost::bind(&AddingImageNodelet::reconfigureCallback, this, _1, _2);
> > +      reconfigure_server_->setCallback(f);
> > +
> > +      pnh_->param("approximate_sync", approximate_sync_, true);
> > +      img_pub_ = advertiseImage(*pnh_, "image", 1);
>
> One question,
> why we should use opencv's gui?
> we can use visualize tool such that image_view.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ros-perception/opencv_apps/pull/29/files/8584daaa82563150482f11d4e889cdc745d18aef#r76930698>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAeG3NfAILJtKKFZVom4JsnVIl54FNpVks5qlR82gaJpZM4Jw22Z>
> .
>